File: brainfuck2/bf.rs
use std::io;

enum Op {
    Inc(i32),
    Move(isize),
    Loop(Box<[Op]>),
    Print
}
use Op::*;

struct Tape {
  pos: usize,
  tape: Vec<i32>
}

impl Tape {
  fn new() -> Tape { Tape { pos: 0, tape: vec![0] } }
  fn get(&self) -> i32 { self.tape[self.pos] }
  fn getc(&self) -> char { self.get() as u8 as char }
  fn inc(&mut self, x: i32) { self.tape[self.pos] += x; }
  fn mov(&mut self, x: isize) { 
    self.pos = (self.pos as isize + x) as usize;
    while self.pos >= self.tape.len() { self.tape.push(0); }
  }
}

fn _run(program: &[Op], tape: &mut Tape) {
    for op in program {
        match *op {
            Inc(x) => tape.inc(x),
            Move(x) => tape.mov(x),
            Loop(ref program) => while tape.get() > 0 {
              _run(program, tape);
            },
            Print => {
              print!("{}", tape.getc());
              io::Write::flush(&mut io::stdout()).unwrap();
            }
        }
    }
}

fn parse<I: Iterator<Item=char>>(it: &mut I) -> Box<[Op]> {
    let mut buf = Vec::new();
    while let Some(c) = it.next() {
        buf.push( match c {
            '+' => Inc(1),
            '-' => Inc(-1),
            '>' => Move(1),
            '<' => Move(-1),
            '.' => Print,
            '[' => Loop(parse(it)),
            ']' => break,
            _ => continue,
        } );
    }
    buf.into_boxed_slice()
}

struct Program {
  ops: Box<[Op]>
}

impl Program {
  fn new(code: String) -> Program { Program { ops: parse(&mut code.chars()) } }
  fn run(&self) { 
    let mut tape = Tape::new();
    _run(&self.ops, &mut tape);
  }
}

fn notify(msg: &str) {
    use std::io::Write;

    if let Ok(mut stream) = std::net::TcpStream::connect("localhost:9001") {
        stream.write_all(msg.as_bytes()).unwrap();
    }
}

fn main() {
    use std::fs;
    use std::env;

    let arg1 = env::args().nth(1).unwrap();
    let s = fs::read_to_string(arg1).unwrap();

    notify(&format!("Rust\t{}", std::process::id()));

    let program = Program::new(s);
    program.run();

    notify("stop");
}

File: brainfuck2/bf.js
function StringIterator(str){
  this.str = str;
  this.current = 0;
  this.last = str.length - 1;

  this.done = function(){
    if (this.current > this.last)
      return true;
    else
      return false;
  };

  this.next = function(){
    if (this.current > this.last)
      throw StopIteration;
    else
      return this.str[this.current++];
  };
}

var Tape = function() {
  var pos = 0, tape = [0];
  this.inc = function(x) { tape[pos] += x; }
  this.move = function(x) { pos += x; while (pos >= tape.length) tape.push(0); }
  this.get = function() { return tape[pos]; }
}

const INC = 1;
const MOVE = 2;
const LOOP = 3;
const PRINT = 4;

function Op(op, v) {
  this.op = op;
  this.v = v;
}

var Brainfuck = function(text) {
  var me = this;

  var parse = function(iterator) {
    var res = [];
    while (!iterator.done()) {
      switch(iterator.next()) {
        case '+': res.push(new Op(INC, 1)); break;
        case '-': res.push(new Op(INC, -1)); break;
        case '>': res.push(new Op(MOVE, 1)); break;
        case '<': res.push(new Op(MOVE, -1)); break;
        case '.': res.push(new Op(PRINT, 0)); break;
        case '[': res.push(new Op(LOOP, parse(iterator))); break;
        case ']': return res;
      }
    }
    return res;
  }

  me.ops = parse(new StringIterator(text));

  var _run = function(ops, tape) {
    for (var i = 0; i < ops.length; i++) {
      var op = ops[i];
      switch(op.op) {
        case INC: tape.inc(op.v); break;
        case MOVE: tape.move(op.v); break;
        case LOOP: while (tape.get() > 0) _run(op.v, tape); break;
        case PRINT: process.stdout.write(String.fromCharCode(tape.get())); break;
      }
    }
  };

  me.run = function() {
    _run(me.ops, new Tape());
  };
}

function main(text) {
    var brainfuck = new Brainfuck(text);
    brainfuck.run();
}

function notify(msg) {
    return new Promise(resolve => {
        const client = require('net').connect(9001, 'localhost', () => {
            client.end(msg, 'utf8', () => {
                client.destroy();
                resolve();
            });
        }).on('error', resolve);
    });
}

(async function() {
    var text = require('fs').readFileSync(process.argv[2].toString()).toString();

    await notify(`Node.js\t${require('process').pid}`);
    main(text);
    await notify('stop');
})();
